application/src/main/java/org/thingsboard/server/service/alarm/rule/DefaultTbAlarmRuleStateService.java

```diff
@@ -0,0 +1,550 @@
+    private void addAlarmRule(TenantId tenantId, AlarmRule alarmRule) {
+        Map<AlarmRuleId, AlarmRule> tenantRules = rules.get(tenantId);
+        if (tenantRules != null) {
+            tenantRules.put(alarmRule.getId(), alarmRule);
+            myEntityStateIds.get(tenantId).stream()
+                    .filter(entityId -> isEntityMatches(tenantId, entityId, alarmRule))
+                    .forEach(entityId -> {
+                                EntityState entityState = entityStates.get(entityId);
+                                myEntityStateIdsPerAlarmRule.computeIfAbsent(alarmRule.getId(), key -> ConcurrentHashMap.newKeySet()).add(entityId);
+                                entityState.addAlarmRule(alarmRule);
+                            }
+                    );
+        }
+    }
+
+    private void deleteAlarmRule(TenantId tenantId, AlarmRuleId alarmRuleId) {
+        Map<AlarmRuleId, AlarmRule> tenantRules = rules.get(tenantId);
+
+        if (tenantRules == null) {
+            return;
+        }
+
+        tenantRules.remove(alarmRuleId);
+
+        Set<EntityId> stateIds = myEntityStateIdsPerAlarmRule.remove(alarmRuleId);
+
+        if (stateIds != null) {
+            stateIds.forEach(id -> {
+                EntityState entityState = entityStates.get(id);
+                Lock lock = entityState.getLock();
+                try {
+                    lock.lock();
+                    entityState.removeAlarmRule(alarmRuleId);
+                    if (entityState.isEmpty()) {
+                        entityStates.remove(entityState.getEntityId());
+                        myEntityStateIds.get(tenantId).remove(entityState.getEntityId());
+                        stateStore.remove(tenantId, id);
+                    }
+                } finally {
+                    lock.unlock();
+                }
+            });
+        }
+    }
+
+    private void deleteTenant(TenantId tenantId) {
+        Map<AlarmRuleId, AlarmRule> tenantRules = rules.get(tenantId);
+        if (tenantRules != null) {
+            tenantRules.keySet().forEach(alarmRuleId -> deleteAlarmRule(tenantId, alarmRuleId));
+            rules.remove(tenantId);
+        }
+    }
+
+    @Override
+    protected void onTbApplicationEvent(PartitionChangeEvent event) {
+        if (!ServiceType.TB_RULE_ENGINE.equals(event.getServiceType())) {
+            return;
+        }
+
+        var internalPartitions = event.getPartitionsMap().entrySet().stream()
+                .filter(entry -> entry.getKey().isInternal())
+                .flatMap(entry -> entry.getValue().stream())
+                .collect(Collectors.toSet());
+
+        Set<TopicPartitionInfo> addedPartitions = diffSets(myPartitions, internalPartitions);
+        Set<TopicPartitionInfo> removedPartitions = diffSets(internalPartitions, myPartitions);
+
+        myPartitions.clear();
+        myPartitions.addAll(internalPartitions);
+
+        log.info("calculated removedPartitions: {}", removedPartitions);
```

Comment:
Use `org.thingsboard.server.service.partition.AbstractPartitionBasedService` to handle partition changes.